### ISSUE TITLE
Audio config possible?

### ISSUE THREAD

**Issue by hoyane** [mdhiggins/sickbeard_mp4_automator#1172]
Hi, is it possible to configure this so that in the end, no matter how many original streams there were, there is only a default 2 ch aac and the original source 5+ ch audio no matter the format?

I have the settings close to that but what messes my automation up is the codec setting under audio. Because no matter how many formats you add there it will convert them to the first one listed. If it simply did a copy on any format listed there i would be golden. I just want a plex audio stream and the original multi no matter how many were there originally.

Would it make sense to add an option for formats to passthrough, separate from the one that converts them to the first listed?

**Comment by mdhiggins** [comment]
If the codec is listed it will copy assuming there’s not some other parameter you set that prevents copying. It will only convert if the codec isn’t on the list 

Feel free to post some example files that aren’t giving you what you want along with your autoProcess.ini settings

Another avenue would be to list aac as your only codec set max channels to 2 and turn on copy-original.

**Comment by mdhiggins** [comment]
Closing this since its just a question but the functionality should be present (in more than 1 way) but feel free to continue commenting with questions

**Comment by hoyane** [comment]
Thanks for the reply. 

Config: https://pastebin.com/raw/zVFf53yN
Output: https://pastebin.com/raw/Y66395Xv

I have two things in this one that are vexing me. First is the error it gets about the experimental dca codec. It says to use the -strict -2 option but it just throws an error when i try to use it. How can I get past this error?

The second issue which is what this post was really about is the tracks i'm getting on output. In this example the original has a aac 2ch, dts 6ch and another aac 2ch. What I would like is to have just one 2ch aac and the original surround track no matter what format it is, dts, ac3, eac3 or whatever it might be. When I process this it wants to use all 3 tracks.

**Comment by JySzE** [comment]
@hoyane 

Try this:

`postopts = -strict,-2`

**Comment by mdhiggins** [comment]
Didn't realize DTS needed strict parameters, I'll add those at the codec level so that's not required to be part of postops

6e89133c745ab95795ebfef84f7f6451924e60a3

To get what you want specifically what I would do is enable universal audio and enable first-stream-of-language. I would also turn on sort-streams and prefer-more-channels

By doing this you'll get priority of your language and 6 channel audio. If you have all the codecs on the whitelist they'll be copied, and a 2 channel option will be made only for that channel, and the rest of the tracks should be ignored.

This should result in a file with a copied 6 channel audio track and a stereo audio track and discard any additional tracks.

There's no specific way to copy a 2 channel and 6 channel and ignore the rest. It would be hard to know which tracks to ignore but that should at least give you the output you're looking for

**Comment by hoyane** [comment]
So far its giving me the streams i want but can I make the 2ch aac the default?

**Comment by mdhiggins** [comment]
It chooses the default (assuming the source had no default) based on the sorting so you kind of run into a problem here in that you're preferring more channels but then don't want to use that stream as the default

**Comment by mdhiggins** [comment]
501c7d1c0191c49667fcd4fb2a2c845afc6a5438

Here, update that splits that option up and there's a default-more-channels option so you can set that to false and that should let you have the 2 channel be default

**Comment by mdhiggins** [comment]
ce39048629293547d1a2f673d1719bb917c892cc

Added a `preserve-source-disposition` option if you want to ignore any preexisting default flags coming from the source file to ensure your sorting gives you a consistent default track

**Comment by JySzE** [comment]
@mdhiggins 

Does `preserve-source-disposition` also preserve the "forced" flag?

**Comment by mdhiggins** [comment]
Yes for subtitles

**Comment by mdhiggins** [comment]
Hm, care to share the logs? Having it false should wipe out any forced or default disposition settings

**Comment by JySzE** [comment]
Log: https://pastebin.com/kixFVbqb

Original File Meta: https://pastebin.com/v72mXDbf

**Comment by mdhiggins** [comment]
So the script will still set dispositions and with only one audio track that will still become the default, and it doesn't look like its setting anything as 'forced' only 'default' (mp4 doesn't really respect forced audio anyway) so appears to be working as intended

What the preserve-source-disposition option does is ignore the sources flags and makes up its own default track based on your sorting, language, and channel preferences. If the source media has chosen another default with that option enabled, it won't override that default (or forced)

**Comment by mdhiggins** [comment]
And just to clarify while the script doesn't attempt to make up a 'forced' track it will still preserve that setting if the option is enabled for both audio and subtitles

**Comment by hoyane** [comment]
Thanks a bunch! I've run it through several files and it appears to be working good.

Now i'm going to bug ya about two small fry inquiries.

First: Is it possible to make it label the audio channels with a simple 2ch or 6ch. It just creates audio 1 and audio 2 when you select audio in a player.

Second: I am in the subtitle section now and wonder if possible to have a first-stream-of-language for the subtitles like there is for audio? Furthermore, I get lots of files with many sub streams and some will not even work. Would it be possible to also have an option that disregards all subs and tries to download new ones?

**Comment by mdhiggins** [comment]
I don't believe FFMPEG can label the tracks

Those subtitle features would need to be built but could be doable

**Comment by mdhiggins** [comment]
536d0721b92fa94189daba9b64120568a79c4242

You're capped for new feature requests now 😝

**Comment by hoyane** [comment]
For how long? lol jk Thanks a lot for adding these things!

**Comment by mdhiggins** [comment]
bbdecce6bd0c16fda6173ed921e9202777ff23f8

Small tweak to prevent situations where a language would get blocked but it wouldn't find a valid codec and then end up not included in the final output

**Comment by hoyane** [comment]
I am having a issue where it is not downloading new subs but it doesn't give any error. Is there a debug switch so i can see more info?

**Comment by mdhiggins** [comment]
you can set your logging level to debug inside config/logging.ini

Also subliminal does an independent scan of the source file to see if subs are already present in a language which I had to cancel out if you're ignore embedded subs

Try this update
05fb45a4426cdfb219ea4f1cbba40306ad2e8eb6

**Comment by hoyane** [comment]
That did it! Thanks

**Comment by hoyane** [comment]
Me again! Not for anything new though. So i started organizing my library and it was naming all my files with DTS or AC3 as the audio label ex: Movie Title (2001) 1080p x264 AC3.mp4. I see in the log automator doesn't actually reorder the streams but only marks the AAC which is the 2nd stream as the default. I know i've asked a lot of late so if you haven't the time no worries I got a good deal out of it already or maybe it can't be done but can this be remedied so the order is actually changed?

**Comment by hoyane** [comment]
As a conversational aside, it looks like ffmpeg can label audio streams.

ffmpeg -i input.mp4 -map 0 -metadata:s:a:0 title="One" \
-metadata:s:a:1 title="Two" output.mp4

**Comment by mdhiggins** [comment]
Good to know, integrating titles should be doable then

**Comment by mdhiggins** [comment]
6105e5b32e8c163c0f30aa44787365b89d72ed34
Titles

**Comment by hoyane** [comment]
Very nice!

**Comment by unsocialsoul** [comment]
Hi, 2 questions, 

1) What setting should i use to disable the Universal Audio?
2) How much additional space would Universal Audio take up?

**Comment by mdhiggins** [comment]
1. Leave codec blank
2. About 10% of original file size

**Comment by unsocialsoul** [comment]
Ah ok. thanks for replying. I assume there wouldnt be any issues with disabling this? In the older versions, i had disabled this, and never faced any issues. Just wondering what are the benefits of having the universal audio? What are the kind of devices that would need this?

I have been using your tool for the past few years, and I absolutely love it. Just wanted to say thanks!

**Comment by mdhiggins** [comment]
Nah it’s an optional feature feel free to disable. 

It’s designed to be a track that minimizes transcoding in plex to portable devices with more limited hardware. 

Glad you’re enjoying!

**Comment by hoyane** [comment]
So unfortunately after loading some converted files into plex, it does not detect the aac as the default audio, instead starting with the first stream. Although playing in something like vlc does detect aac as default. So Plex only looks at the order of the streams i suppose. The same goes for sonarr/radarr when renaming includes the audio type. 

For me the perfect plex file has a 2ch aac as the primary/default stream which plays perfect in every notable device I can think of. Then it has a copy of the surround source in case i'm on a device that has surround capabilities as the second stream that I can switch to. Everything else is stripped away just to make it tidy and smaller. 

Do you think this can be done?

**Comment by mdhiggins** [comment]
Yeah this is doable, use the following settings
```
sort-streams = True
prefer-more-channels = False
default-more-channels = False
```

**Comment by hoyane** [comment]
So my input has a 6ch and a 2ch stream. My current config gives me only 1 aac output stream.
The only difference between this and my previous config is prefer-more-channels was true. My previous config had me getting a 6ch + 2ch output but they were in the wrong order. Some programs don't care that the 2ch is marked default and goes with the first stream in the file. Except vlc, when i played it in there it actually started with the 2ch. But everything from filebot to sonarr/radarr to plex will see that the dts is the first stream and identify by that instead of the aac. The image below is from vlc. Stream 1 needs to be the aac for this to work.

![image](https://user-images.githubusercontent.com/43273937/77253265-f436a380-6c26-11ea-8294-9c03a2636143.png)

config: https://pastebin.com/raw/Sa0xJkFm
output: https://pastebin.com/raw/VeE5bhNf

**Comment by mdhiggins** [comment]
This is because of `first-stream-of-language = True`

Try setting prefer-more-channels back to to True but leave default-more-channels False

**Comment by hoyane** [comment]
With that i get the two streams but they are in the wrong order. filebot, sonarr and plex will all ignore the 2ch default and use the first stream instead. For the below image i need the aac to be on top so its the first stream. Seems strange that ffmpeg would look at stream order and default setting differently.

![image](https://user-images.githubusercontent.com/43273937/77254041-84c3b280-6c2c-11ea-9a26-a85c3800bf1c.png)

**Comment by mdhiggins** [comment]
pushed an update that should address this

**Comment by hoyane** [comment]
Thank you! That worked nicely. A question for you about performance. When i run a process where the video is copy and the audio does what we've done here, 2ch create and 6ch copy. When I view my system monitor it shows no hit more than 10% at any time. That's on cpu and disk usage. So is ffmpeg a case of hardware never holding it back but the software itself is the bottleneck? Or should i be seeing more of a system resource hit than that? During the moov process i get 40-50%  hdd utilization. With what it's doing I would anticipate a 10gb file being done in about 5 minutes but these take more like 10 minutes to finish.

**Comment by mdhiggins** [comment]
This is an FFMPEG issue. Not all encoders have good utilization or multiple cores or even single core. Sometimes you are HDD limited as well.

**Comment by hoyane** [comment]
I was running all this on windows and decided to load up a linux version for testing. However when i run the script i get this....  https://pastebin.com/raw/XepiK2Q0

Can you tell me what that error is about?

**Comment by mdhiggins** [comment]
This looks like a failure with sysloghandler. Not sure why it would fail but its really more trouble than its worth since I tried adding that
Just put up a new patch that removes that see if things are fixed
546db1f82dbc1f573a2d498686ce7194f7e1f024

**Comment by hoyane** [comment]
That worked thank you.

**Comment by hoyane** [comment]
I'm back!
Now the issue is with nzbget. automator works fine manually but when used between radarr and nzbget i get this error.

![image](https://user-images.githubusercontent.com/43273937/77427244-0be76680-6da4-11ea-9ebf-97121d2830f1.png)

I've triple checked my pathing, googled with little results and added full path to python at the start of the py scripts. These must be Windows quirks cause I never had this much trouble setting up on linux in the past.

**Comment by mdhiggins** [comment]
Yeah that's a tough one, it seems like python wasn't even able to launch which limits by debugging options
What changes did you make to the file?

**Comment by mdhiggins** [comment]
https://docs.python.org/3/using/windows.html#python-launcher-for-windows

Might be worth looking into that, that seems like that needs to be setup to translate the linux shebang lines into windows specific python installations

**Comment by hoyane** [comment]
The only changes i made to any file is the full path in the py files for nzbget and postradarr.

**Comment by hoyane** [comment]
I had the launcher installed already. Is python not launching or is the filename not being passed? Based on this line in below image i would almost say that the video filename is not being passed. Also the part that says the path to python.exe is being interpreted by the sript i beliwve, cause right now the only place i have the hard path is in the radarr connection module and it's not coming from that cause i tested by changing that path.

![image](https://user-images.githubusercontent.com/43273937/77432621-4fde6980-6dac-11ea-8ed2-2c472ea2e6fa.png)

**Comment by mdhiggins** [comment]
No such file or directory I believe is coming from the python executable not finding the file. For some reason it looks like its only passing `L` as the filename

![image](https://user-images.githubusercontent.com/3608298/77445221-2f6dd980-6dc3-11ea-8abb-f354104b6e14.png)

**Comment by mdhiggins** [comment]
I think you might have broken your shebang

**Comment by hoyane** [comment]
So right now with that last error the py files are back to the way they were. Since adding the hard path didnt work i removed it. So as it stands nothing is modified at this point. Below is a link to the nzbget debug log. 

https://pastebin.com/raw/xipiXSaL

**Comment by mdhiggins** [comment]
Yeah so again NZBGet isn't  even executing the script
Its executing

```
Tue Mar 24 11:30:36 2020	8136	4820	DEBUG	Command: "C:\Windows\py.exe" "%L" %* (Script.cpp:285:ScriptController::PrepareArgs)
Tue Mar 24 11:30:36 2020	8136	4820	DEBUG	CmdLine: "C:\Windows\py.exe" "L" (Script.cpp:292:ScriptController::PrepareArgs)
Tue Mar 24 11:30:36 2020	8136	4820	DEBUG	Starting process: "C:\Windows\py.exe" "L" (Script.cpp:476:ScriptController::StartProcess)
```

Why its doing that I can't say. This has not begun to touch any of my code NZBGet is executing something incorrectly

**Comment by mdhiggins** [comment]
https://forum.nzbget.net/viewtopic.php?t=1267

https://forum.nzbget.net/viewtopic.php?t=3028

Looks like its an NZBGet issue

You can probably fix it by using shellOverride

NZBGet > Settings > Extension Scripts > Shell Override

.py=C:\path\to\python3.exe

**Comment by hoyane** [comment]
You're awesome! Works now! Sonarr v3 setup worked too!

Thanks

**Comment by hoyane** [comment]
I just can't catch a break! I'm sorry for all the needed help. This was so much less trouble on linux.

I noticed my output didnt have a poster cover and when I checked properties it was not tagged with metadata. It does this on auto and manual conversion. What's strange is it was not doing this previously. I ran 100 movies through no problem.

https://pastebin.com/raw/GKiXW4jw

**Comment by hoyane** [comment]
After looking at that log harder i realize there is no useful info in there. Those are just debug messages not errors. None the less, the files are not tagging when they did before.

**Comment by hoyane** [comment]
I do notice that -metadata:g was not there before.

Postopts
[
    "-threads",
    "0",
    "-metadata:g",
    "encoding_tool=SMA"
]

**Comment by mdhiggins** [comment]
The metadata g think is unrelated. The logs you attached aren't helpful, those are just the postprocess logs tagging happens before that step so those logs are too late. For manual conversion can you post the whole log and I can take a look

**Comment by hoyane** [comment]
I was just in the process of doing just that. Here you go.

https://pastebin.com/raw/UHDBm8gm

**Comment by mdhiggins** [comment]
https://www.themoviedb.org/tv/61018-whose-line-is-it-anyway/seasons

Looks like TMDB is out of date for that show, its failing because it only has data up to season 8

**Comment by hoyane** [comment]
I know TMDB has tv shows also but why does it use that instead of TVDB? The latter has all 15 seasons listed.

**Comment by mdhiggins** [comment]
TVDB's API is garbage and was dumped due to lots of problems. TMDB has about 95% of what TVDB has and you can always add things

**Comment by hoyane** [comment]
When I process this file it appears to be getting processed twice. This is shown in the log and you can see the second one is from the D drive after sonarr moves it. So on the second time in the log it looks like it only copies the aac. However the eac3 does in fact still come out. Another thing is it leaves a srt from processing it a 2nd time on D drive.

Config: https://pastebin.com/raw/A6B8wA0v
Log: https://pastebin.com/raw/VKzNeDEp

What came out:
![image](https://user-images.githubusercontent.com/43273937/77579441-58b46580-6ea8-11ea-8b0b-3b6f3696d8a3.png)


Also with this type file, where source is only a 2ch eac3 the label for the eac3 copy just says English - [English]. So I realized if you fix that then both labels would read the same thing with no differentiation. Would it be possible to add the codec into the label?
AAC Stereo
EAC3 Stereo

![image](https://user-images.githubusercontent.com/43273937/77582920-2c034c80-6eae-11ea-9584-861945f35dee.png)

**Comment by mdhiggins** [comment]
This is converting twice because you have force convert turned on. If you turn that off and ideally don’t use process same extensions this should work as it should.

**Comment by hoyane** [comment]
If i turn process same extensions off does that mean it would not process any mp4 files? Cause I would still want those to go through the process.

**Comment by mdhiggins** [comment]
Correct. You should be ok if you expect MP4 input files to leave it on then but turn off force-convert if you’re going to use settings that duplicate audio tracks like universal audio or copy original. There is logic built into the script to prevent duplicate tracks but that gets forced off by turning on force convert.

**Comment by hoyane** [comment]
Actually I went to change the force convert and it is already set to false.

**Comment by mdhiggins** [comment]
- INFO - Input and output extensions match and every codec is copy, this file probably doesn't need conversion, but conversion being forced [force-convert: True].

Logs seem to indicate it was on for this scenario

**Comment by hoyane** [comment]
I'm sorry you are right. I continue to troubleshoot after asking original question. I had my config file backed up and that was set to force convert. However the config i am on now is set to false and it still does it. Here is the log file while set to false.

Log: https://pastebin.com/raw/9mTz9fF0

**Comment by mdhiggins** [comment]
```
INFO - Input and output extensions match and the file appears to have already been processed by SMA, enable force-convert to override [force-convert: False].
2020-03-25 22:42:44 - resources.mediaprocessor - INFO - Tagging D:\lab\Tv_Shows\Cheers\Season 3\Cheers - S03E16 - Teacher's Pet - WEBDL-1080p h264 AAC.mp4 with TMDB ID 141.
2020-03-25 22:42:44 - resources.metadata - INFO - Tagging file: D:\lab\Tv_Shows\Cheers\Season 3\Cheers - S03E16 - Teacher's Pet - WEBDL-1080p h264 AAC.mp4.
```

Looks like it’s working as intended. It recognizes that the file has already been converted and then proceeds to tagging without passing it through FFMPEG again.

**Comment by hoyane** [comment]
Maybe I was never paying attention in the past but I don't remember it ever looking like it went through twice. I've watched so many manual conversions and never noticed that before. So a couple things about that. Why does it download subtitles twice and then leave them behind externally the 2nd time? 

Also, shouldn't all processing for a file take place in one place before going to final destination? My scratch drive is fast but final destination is not. The idea is to run as much as I can in memdisk to save writes on my drives.

**Comment by mdhiggins** [comment]
Tagging metadata isn’t available until Sonarr/Radarr have control of the file (NZBGet has no metadata information) which is why that’s reserved until the final step. I can try and look into ways to prevent the downloaded sub from happening twice though. But tagging is a very minimal process not writing lots of data, though the QTFastStart component does move lots of data. Unfortunately you can run into playback issues if you try to do QTFastStart before you write tags in some scenarios From: hoyaneSent: Thursday, March 26, 2020 8:30 AMTo: mdhiggins/sickbeard_mp4_automatorCc: Michael Higgins; MentionSubject: Re: [mdhiggins/sickbeard_mp4_automator] Audio config possible? (#1172) Maybe I was never paying attention in the past but I don't remember it ever looking like it went through twice. I've watched so many manual conversions and never noticed that before. So a couple things about that. Why does it download subtitles twice and then leave them behind externally the 2nd time?Also, shouldn't all processing for a file take place in one place before going to final destination? My scratch drive is fast but final destination is not. The idea is to run as much as I can in memdisk to save writes on my drives.—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or unsubscribe.

**Comment by hoyane** [comment]
I understand sonarr has to have it first for the metadata but it cant do that part before moving to final destination?

**Comment by mdhiggins** [comment]
Unfortunately no, Sonarr and Radarr move things, the script isn't activated until after that part is completed

**Comment by hoyane** [comment]
Found two quirks with what we've worked on. 

Do I understand the default-language when it encounters no eng but does find a undefined stream it was assume and mark that stream as english? Or it will just accept it as undefined without labeling it? If the prior then it is not labeling the language on output. I sort of assume the later as it could accidentally label a undefined german and label it as eng.

Second, In this scenario it did not label the audio stream as Stereo.

![image](https://user-images.githubusercontent.com/43273937/77796968-7c54e880-702d-11ea-8f64-9bf68ab9770b.png)

Config: https://pastebin.com/raw/3FaNg9n7
Log: https://pastebin.com/raw/QGiSMz88

**Comment by mdhiggins** [comment]
Looking at the logs that file was never converted because it’s already in the same format and all the codecs were copy so it bypasses the conversion process to save resources.

Default-language assumes any unlabeled languages are what you have default set to. Again you’re not seeing this work because your file wasn’t processed by FFMPEG

I think I might remove that feature though in favor of the new option that checks the encoder to see if it was already processed by SMA as the sole conversion bypass

**Comment by mdhiggins** [comment]
9cd51d17ae4e84939a533f1a6f7e50492bc963fa
See if that helps

**Comment by hoyane** [comment]
This did correctly label the audio stream. While running some things it looks like you changed it to not run the first conversion it was doing in place and now nzbget calls sonarr as soon as its unpacked. So it looks like sonarr moves to final destination and renames it and then runs its post script to convert it. Did you do this to get the metadata before processing?

Problem is that it doesn't do the rename the 2nd time. So if the first time was a ac3 audio then that's in the new name then after convert it is aac but it does not rename it again. Can you add another rename trigger?

1. Also notice that when running multiple nzb since nzbget used to trigger the next only as they finished and now they finish right away they start to stack and process at the same time. Like when one was doing the qtfs another came in and started the conversion at the same time. I suppose in some scenarios where they come in fast this could cause a log jam with too many heavy i/o ops. 

2. Another thing that comes to mind is people that have things happen after a file appears somewhere. Like I used to run a union and when a file would appear it would move it from the first layer location to the second layer location but if it gets converted from mkv to mp4 in the middle of xfer that could gum things up.

3. Also its sort of become a blind process now that you cant see nzbget working on it. Not the end of the world, just saying.

I'm not asking you to address either of these last three. I just want you to be aware since you cant test every scenario.

**Comment by mdhiggins** [comment]
The only chance I made was that conversion won’t be bypassed if all streams are copied which should do the tagging.

The streams are named if there’s a conversion. If you’re not seeing renaming then the conversion isn’t happening but in that case you also shouldn’t be seeing those resulting streams even being created. I’m gonna adjust the logging so it’s less confusing because right now it shows you what it intended to create then blocks. You really don’t want two conversions happening to the same file there’s no point to that. 

For your other points
1. Python logging should be able to handle it
2. This is why things like completed download handling need to be turned off if you’re using sonarr / Radarr AND post downloaded scripts but should be accounted for
3. NZBGet doesn’t have a way to report progress back sadly I’ve wanted this for a while but it’s not supported

**Comment by hoyane** [comment]
In this case i'm not talking about the streams. I mean the actual final filename. I started with a mkv with ac3 audio. When nzbget triggered sonarr it moved and renamed it according to the name scheme i set in sonarr. So it was then MOVIE.NAME.X264.AC3.mkv. Then the script converts it to mp4 but it keeps the same name MOVIE.NAME.X264.AC3.mp4. It should now be named with AAC since it changed.

**Comment by mdhiggins** [comment]
Oh gotcha. The script doesn’t do any renaming of the file that’s all sonarr/radarr you’d have to adjust your renaming in those programs.

**Comment by hoyane** [comment]
Yeah but you cant make the script trigger a rescan/rename as the last step so that it is not outside of the automatic process. If i set force convert will it do it twice like before so that it is already aac before it gets to sonarr? Even if its causing more writes i would rather correct naming be done as part of the auto process.

**Comment by mdhiggins** [comment]
If you’re using an additional converter script from the downloaded the conversion will be done already by the time it gets there. You could also just not include the codec in the final name and not have to worry about this. The script does already trigger a rescan but not another rename. I’m also not sure if Sonarr / Radarr renames will pull codecs for the name or not anyway.

**Comment by mdhiggins** [comment]
I stand corrected, sonarr/radarr can in fact rename based on codecs

New update triggers a rename after everything is done which should fix your issue

**Comment by hoyane** [comment]
Yes but only if you have analyze video enabled. Which i do as i've always had my codecs in teh filename. lol I was just going to abandon that naming after your last post so thanks for following up on that.

**Comment by hoyane** [comment]
Thanks again for making these changes that you have made. I have one  more for you. If you don't think its a good idea then skip it. As you know all my files will have the 2ch aac and then the source file if it was converted. So now i have these files where the audio selection is the same cause the source was also Stereo (see pic). Could you add the codec into the title so you know what your selecting?

AAC Stereo
AC3 Stereo
DTS 5.1

![image](https://user-images.githubusercontent.com/43273937/77923920-b1994a80-7268-11ea-93e1-e61b03c97286.png)

**Comment by mdhiggins** [comment]
I intentionally don't include the codec name because most players show you the codec the stream you're choosing. Additionally, there's lots of codecs with long ugly names

**Comment by hoyane** [comment]
Understood! I want to circle back to something i asked about before regarding a backlog of files to process. The image below is from a different server I use that is still using an older version of automator. That's why they are stacked up in nzbget but it demonstrates how many can stack before the first is finished. The question is will sonarr only process one file at a time and not process the next request until done or would it be trying to run 20 instances at once?

![image](https://user-images.githubusercontent.com/43273937/77973361-8854da80-72b9-11ea-8033-0e4a442dc937.png)

**Comment by mdhiggins** [comment]
That’s more of a Sonarr Radarr question not specific to my script. But I believe it does not queue though I can’t say 100%

**Comment by hoyane** [comment]
I also believe it will not queue. I definitely don't mean this to sound any way other than kosher. i understand this is a project for you that you maybe don't get many dimes for the work you put into it and I am grateful. And i'm def not saying this is your responsibility to fix. So you def don't deserve any grief. That said, I came along and asked you to modify the program so that it could accomplish what I was looking for. As a result of those modifications it has created a scenario no one envisioned of 20 conversions happening at once. 

Again, it's really not my intention to aggravate you but your changes at my behest did change the way the system works. Now if you just cant see any way to resolve that new behavior I will get on the sonarr git page and make a feature request although I would have to get radarr to do this as well since they are separate forks.

If you still disagree that it's on sonarr now I will accept that. i just need to be sure that you hear me is all.

Thanks again for all you do on this project!

**Comment by mdhiggins** [comment]
No hostility intended, just busy so answers are brief. Because the script is sort of independent and without central storage or persistence there's really no good solution to the queueing thing without adding a ton of complexity which would be better suited for the various programs it integrates with

That being said Sonarr/Radarr might actually queue. It definitely has a queueing system I've just honestly never tested it but this would be something that if not implemented would be on them to add. Its easy enough to see if it has one though. Just do a manual import of a few files and watch the folder and see

**Comment by hoyane** [comment]
Yeah unfortunately it looks like it does not. The 0 size duplicate files are the conversions happening at the same time. With repacks or just audio conversions the hit is not so hard but I can imagine 10 video encodes happening at once which could get heavy for a system.

![image](https://user-images.githubusercontent.com/43273937/77979010-1ab0aa80-72c9-11ea-9664-7dcb2a5aff23.png)

**Comment by johnwchambers** [comment]
@hoyane 
Here is an example of how I keep deluge from only converting one thing at a time, notice the 2nd line in the bash script. I pulled that from 'man flock', it's listed as an example that you can put at the top of any bash script pretty much to only allow one instance of it at a time to run while the rest are queued up until the previous one exits. 

Hopefully this helps you out and gets yous running to where things are queued the way you want them to be. So far this has been working great for my setup.

>
    #!/bin/sh
    [ "${FLOCKER}" != "$0" ] && exec env FLOCKER="$0" flock -x "$0" "$0" "$@" || :

        torrentid=$1
        torrentname=$2
        torrentpath=$3

        pathextraction=/dwnldr/tor/extracted
        pathconversion=/dwnldr/tor/converted

        if [[ -d "$torrentpath/$torrentname" ]]; then
            #echo "$torrentpath/$torrentname" + " is a directory"
            mkdir "$pathextraction/$torrentname"
            mkdir "$pathconversion/$torrentname"
            /opt/deluge/unrarall/unrarall -v -o "$pathextraction/$torrentname" "$torrentpath/$torrentname"
            extracted=$?
            wait $!
            if [ $extracted -eq 0 ]; then
                if [[ -n $(find "$pathextraction/$torrentname" -type f -size +50M) ]]; then
                    python3 /opt/mp4automator/delugePostProcess.py "$torrentid" "$torrentname" "$pathextraction/$torrentname" "forcepath"
                else
                    python3 /opt/mp4automator/delugePostProcess.py "$torrentid" "$torrentname" "$torrentpath"
                fi
            else
                python3 /opt/mp4automator/delugePostProcess.py "$torrentid" "$torrentname" "$torrentpath"
            fi
            rm -rf "$pathextraction/$torrentname"
        elif [[ -f "$torrentpath/$torrentname" ]]; then
            #echo "$torrentpath/$torrentname" + " is a file"
            python3 /opt/mp4automator/delugePostProcess.py "$torrentid" "$torrentname" "$torrentpath"
        fi

        exit 0

So for my postSonarr bash script, it's the following:

>
    #!/bin/sh
    [ "${FLOCKER}" != "$0" ] && exec env FLOCKER="$0" flock -x "$0" "$0" "$@" || :

        /usr/bin/python3 /opt/mp4automator/postSonarr.py
        exit 0

And for my postRadarr bash script, it's the following:

>
    #!/bin/sh
    [ "${FLOCKER}" != "$0" ] && exec env FLOCKER="$0" flock -x "$0" "$0" "$@" || :

        /usr/bin/python3 /opt/mp4automator/postRadarr.py
        exit 0

**Comment by hoyane** [comment]
Im curious! What kind of metadata does it get from sonarr that it has run automator at sonarr stage? This metadata cant be grabbed from the sonarr api queue while still in nzbget in order to utilize nzbget existing ability to queue file to automator?

**Comment by hoyane** [comment]
@johnwchambers Thanks for that.  It will come in handy on the linux setup i have but I don't think it will work out of a windows .bat file. Unless its python code. Is it?

**Comment by hoyane** [comment]
@mdhiggins as an aside that I don't really care about but will note that the script did not run when I imported existing files from hdd. I had to add a show to sonarr and let it do it's thing to get that screenshot above. I did have the on import box in sonarr checked also.

**Comment by johnwchambers** [comment]
> Im curious! What kind of metadata does it get from sonarr that it has run automator at sonarr stage? This metadata cant be grabbed from the sonarr api queue while still in nzbget in order to utilize nzbget existing ability to queue file to automator?

This is my workflow:
- NZBGet/Deluge downloads file. NZBGetPostProcess/delugePostProcess scripts are executed to convert the content and then converted files are passed to Sonarr/Radarr.
- Sonarr/Radarr configured with post*arr scripts. Upon import/scanning, those are executed.
- The post*arr process does the metadata tagging of the files and optimizes them for streaming. Places file in proper location and refreshes Plex library.

So bottom line is, post *arr scripts are required for metadata tagging of converted files. The post downloader processes of SMA do not do any metadata tagging of the converted content, that all happens at the multimedia management application.

> @johnwchambers Thanks for that. It will come in handy on the linux setup i have but I don't think it will work out of a windows .bat file. Unless its python code. Is it?

The flock command is part of the linux bash shell. I'm not sure if there is an equivalent of it that you can run on a windows platform.

**Comment by mdhiggins** [comment]
Using completed download handling radarr/sonarr do seem to queue for me, not files don't start moving until the post script of the previous has finished

**Comment by mdhiggins** [comment]
@hoyane Manual import should trigger the script, I use this all the time to test encoding and to add media to my library but I think bulk imports / rescan meant to import existing media does not trigger this. Not sure what you mean by the `on import` box though

**Comment by hoyane** [comment]
Ahh this is probably cause I didn't realize you need completed download
handling on now.

**Comment by mdhiggins** [comment]
Can't remember specifically your setup but don't use completed download handling with a post downloader script for the downloader (IE don't use completed download handling and SABPostProcess). They're meant to be exclusive. That being said the scan that's triggered by the post scripts should also be queueing via the same system. The only one that might not queue would be a bulk import job though I haven't tested that
